spark/src/main/scala/geotrellis/spark/io/FilteringLayerReader.scala

```diff
@@ -1,17 +1,26 @@
 package geotrellis.spark.io
 
 import geotrellis.spark.Boundable
+import geotrellis.spark.io.index.KeyIndex
+import geotrellis.spark.io.json._
+import spray.json.JsonFormat
 
-abstract class FilteringLayerReader[ID, K: Boundable, M, ReturnType] extends LayerReader[ID, ReturnType] {
+abstract class FilteringLayerReader[ID, K: Boundable, M, ReturnType] extends LayerReader[ID, K, ReturnType] {
 
-  def read(id: ID, rasterQuery: RDDQuery[K, M], numPartitions: Int): ReturnType
+  def read[I <: KeyIndex[K]: JsonFormat](id: ID, rasterQuery: RDDQuery[K, M], numPartitions: Int, format: JsonFormat[I]): ReturnType
+
+  def read[I <: KeyIndex[K]: JsonFormat](id: ID, numPartitions: Int, format: JsonFormat[I]): ReturnType =
```

Comment:
We should know how to deserialise the `KeyIndex`, and we need to know its type. But we can use generic `KeyIndex[K]` for in-built types (look at the json format for the `KeyIndex[K]` type).